### PR TITLE
Refactor work loop

### PR DESCRIPTION
**Pull request by marcellp**
# Background
#### Links to any Pull Requests related to this
https://github.com/biomage-ltd/api/pull/210

#### Anything else the reviewers should know about the changes here

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [ ] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [ ] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: python/src/worker/config/__init__.py [hunk 1]
```diff
@@ -55,14 +56,27 @@ def get_label(self, label_key, default=None):
...
+    @property
+    def REDIS_CLIENT(self):
+        try:
+            return self.redis
+        except AttributeError:
+            if cluster_env == "development" or cluster_env == "test":
+                self.redis = redis.Redis(host="host.docker.internal", port=6379)
+            else:
+                self.redis = redis.Redis(
+                    host="master.biomage-redis-staging.aykd0e.euw1.cache.amazonaws.com",
```
Comment (marcellp) [inline python/src/worker/config/__init__.py:72]: this needs to be dynamically fetched from aws

File: python/src/worker/tasks/cluster_cells.py [hunk 2]
```diff
@@ -41,7 +43,33 @@ def _format_result(self, raw):
...
+                {
+                    "$prepend": cell_set_object,
+                }
+            ],
+        )
+
+        info(r.status_code)
+    
+    def _format_result(self, cell_set_object):
+        return [Result(json.dumps(cell_set_object))]
```
Comment (marcellp) [inline python/src/worker/tasks/cluster_cells.py]: Set `cacheable` to be `False` for clustering to always run (needed for `update_through_api`).

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -196,18 +196,18 @@ jobs:
...
+      #   run: |-
+      #     pip install pytest-cov
+      #     ci_env=`bash <(curl -s https://codecov.io/env)`
+
+      #     docker run -v /home/runner/work/worker/worker/data:/data $ci_env -e CI=true --env CLUSTER_ENV=test --net="host" --entrypoint /usr/bin/env $PYTHON_IMAGE_NAME /bin/bash -c "
+      #       python3 -m pytest --cov=. --cov-report=xml
+      #       bash <(curl -s https://codecov.io/bash)
+      #       "
+      #   env:
+      #     PYTHON_IMAGE_NAME: ${{ format('{0}/{1}:{2}-python-ci', steps.login-ecr.outputs.registry, needs.build-docker.outputs.repo-name, needs.build-docker.outputs.tag) }}
```
Comment (ivababukova) [inline .github/workflows/ci.yaml:210]: we need to enable this again

File: python/src/worker/response.py [hunk 4]
```diff
@@ -61,58 +56,67 @@ def _upload(self, result):
...
+                        'Key': 'experimentId',
+                        'Value': self.request['experimentId']
+                    },
+                    {
+                        'Key': 'requestType',
+                        'Value': self.request['body']['name']
+                    },
+
+                    # TODO: this needs to be removed and a proper
+                    # ACL system implemented before merging.
```
Comment (ivababukova) [inline python/src/worker/response.py]: this comment is inaccurate -- we will be implementing a proper ACL system after merging, most likely in the next sprint

File: python/src/worker/result.py [hunk 5]
```diff
@@ -16,18 +16,14 @@ def __init__(
         self.error = error
         self.cacheable = cacheable
 
-    def get_result_object(self, resp_format=False, s3_path=None):
+    def get_result_object(self, resp_format=False):
```
Comment (ivababukova) [inline python/src/worker/result.py:19]: do we actually use this `resp_format` for anything?

File: python/src/worker/tasks/__init__.py [hunk 6]
```diff
@@ -6,6 +6,8 @@ class Task(ABC):
 
     def __init__(self, msg):
         self.task_def = msg["body"]
+        self.request = msg
```
Comment (ivababukova) [inline python/src/worker/tasks/__init__.py]: looks like this `self.request` is only used in the cluster cells task. If that is indeed the case, wouldn't it be better to move this `self.request` to be a variable only for the `ClusterCells` class and add it in its init method?

File: python/src/worker/response.py (hunk 4 above)
Comment (marcellp) [reply python/src/worker/response.py]: fixed

File: python/src/worker/tasks/__init__.py (hunk 6 above)
Comment (marcellp) [reply python/src/worker/tasks/__init__.py]: hmm, i seem to recall there was an issue with it but it seems to be working. i'll put it back there and see

File: python/src/worker/result.py (hunk 5 above)
Comment (marcellp) [reply python/src/worker/result.py:19]: yes, we use it, when `resp_format` is `True`, the result object is formatted to be uploaded to S3. When it is set to `False`, it is formatted to be sent to the UI (i.e. no inlining is done for the results)

File: python/src/worker/config/__init__.py [hunk 7]
```diff
@@ -88,12 +102,16 @@ def QUEUE_NAME(self):
     LOCAL_DIR=os.path.join(os.pardir, os.pardir, "data"),
 )
 
+config.API_URL=f"http://api-{config.SANDBOX_ID}.api-{config.SANDBOX_ID}.svc.cluster.local:3000"
```
Comment (kafkasl) [inline python/src/worker/config/__init__.py:105]: any reason it's not included in the `Config(...)` constructor?

File: python/src/worker/response.py [hunk 8]
```diff
@@ -61,58 +56,67 @@ def _upload(self, result):
         if was_enabled:
             xray.global_sdk_config.set_sdk_enabled(False)
 
-        client.put_object(Key=key, Bucket=self.s3_bucket, Body=body)
+        client.put_object(Key=ETag, Bucket=self.s3_bucket, Body=body)
```
Comment (kafkasl) [inline python/src/worker/response.py:59]: aren't ETags susceptible to change for the same object if it's multipart of different regions or stuff like that? is it ok to use them as caching keys? or the results are just retrieved using the tags?

File: python/src/worker/response.py [hunk 9]
```diff
@@ -61,58 +56,67 @@ def _upload(self, result):
...
+
+        client.put_object_tagging(
```
Comment (kafkasl) [inline python/src/worker/response.py:61]: is this tagging used to find the corresponding files? the name is for for easier debug?

File: python/src/worker/response.py [hunk 10]
```diff
@@ -61,58 +56,67 @@ def _upload(self, result):
...
-            MessageAttributes={
-                "type": {"DataType": "String", "StringValue": "WorkResponse"}
-            },
-            MessageStructure="json",
-        )
-
-        info(f"Request with UUID {self.request['uuid']} successfully sent to SNS, message ID: {r['MessageId']}.")
+            io.Emit(
+                f'{self.request["experimentId"]}-{self.request["body"]["name"]}',
+                self._construct_response_msg(brief=True)
```
Comment (kafkasl) [inline python/src/worker/response.py:100]: I understand that brief messages don't have their results uploaded into S3, is this stuff like gene lists?

File: python/src/worker/response.py [hunk 11]
```diff
@@ -61,58 +56,67 @@ def _upload(self, result):
...
         if was_enabled:
             xray.global_sdk_config.set_sdk_enabled(True)
 
-        return key
+        return ETag
 
-    def _send_notification(self, mssg):
-        sns = boto3.client("sns", **config.BOTO_RESOURCE_KWARGS)
+    def _send_notification(self):
+        io = Emitter({"client": config.REDIS_CLIENT})
```
Comment (kafkasl) [inline python/src/worker/response.py:93]: if redis fails then the worker won't be able to notify anything back right? I assume this is ok in the case where the results are uploaded into S3 as the UI will still check if new results are available as a fallback. however, what would happen for brief messages? they'd be simply lost?

File: python/src/worker/tasks/cluster_cells.py [hunk 12]
```diff
@@ -41,7 +44,33 @@ def _format_result(self, raw):
...
             )
-        return [Result(json.dumps(cell_set), cacheable=False)]
+        
+        return cell_set
+    
+    def _update_through_api(self, cell_set_object):
+        r = requests.patch(
+            f"{config.API_URL}/v1/experiments/{self.request['experimentId']}/cellSets",
+            headers={
+                "content-type": "application/boschni-json-merger+json",
```
Comment (kafkasl) [inline python/src/worker/tasks/cluster_cells.py:54]: are we using this custom content type to be able to merge directly objects (the cell sets) with the json (the stored celll sets) ?

File: python/src/worker/tasks/cluster_cells.py [hunk 13]
```diff
@@ -71,4 +100,9 @@ def compute(self):
         resR = pd.DataFrame(resR)
         resR.set_index("_row", inplace=True)
         resR["cluster"] = pd.Categorical(resR.cluster)
-        return self._format_result(resR)
+
+        # Convert it into a JSON format and patch the API directly
```
Comment (kafkasl) [inline python/src/worker/tasks/cluster_cells.py:104]: I assume it's actually an object but in JSON format?

File: python/src/worker/response.py (hunk 8 above)
Comment (marcellp) [reply python/src/worker/response.py:59]: this is an ETag in the traditional sense, not in the S3 sense. it is confusing, but the AWS term is misleading.

ETags are basically hashes that identify whether a certain resource needs to be refetched, see https://en.m.wikipedia.org/wiki/HTTP_ETag

File: python/src/worker/response.py (hunk 9 above)
Comment (marcellp) [reply python/src/worker/response.py:61]: it's for all of that, but also for setting up custom retention policies for certain types of work (like differential expression) as well as for access control in the future

File: python/src/worker/response.py (hunk 10 above)
Comment (marcellp) [reply python/src/worker/response.py:100]: brief just means the type of result is the one that is sent to the UI (ie does not have results, only whether or not there was an error and if the result is cacheable)

File: python/src/worker/tasks/cluster_cells.py (hunk 13 above)
Comment (marcellp) [reply python/src/worker/tasks/cluster_cells.py:104]: we convert python dictionary into json and add it as part of the json patch

File: python/src/worker/tasks/cluster_cells.py (hunk 12 above)
Comment (marcellp) [reply python/src/worker/tasks/cluster_cells.py:54]: yes, it's a way to update just part of the cell sets, ie the clusters, irrespective of the content the user produced such as custom annotations

File: python/src/worker/response.py (hunk 11 above)
Comment (marcellp) [reply python/src/worker/response.py:93]: you misunderstood brief, see above

File: python/src/worker/config/__init__.py (hunk 7 above)
Comment (marcellp) [reply python/src/worker/config/__init__.py:105]: yes, you can't access members of config in the initializer of config

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (marcellp) [reply .github/workflows/ci.yaml:210]: we decided not to enable it and fix it as a separate ticket along with all the other failing worker tests

File: python/src/worker/config/__init__.py (hunk 1 above)
Comment (marcellp) [reply python/src/worker/config/__init__.py:72]: we will fix it as part of a ticket

**Review by cosa65 — APPROVED** [review]

File: python/src/worker/consume_message.py [hunk 14]
```diff
@@ -97,11 +109,20 @@ def consume():
 
     if timeout <= datetime.datetime.utcnow():
         info(
-            f"Skipping processing task with uuid {mssg_body['uuid']}"
-            f"{mssg_body['uuid']} as its timeout of {timeout} has expired..."
+            f"Skipping processing task with ETag {mssg_body['ETag']} "
+            f"as its timeout of {timeout} has expired..."
```
Comment (ogibson) [inline python/src/worker/consume_message.py:113]: Why do we still have timeouts in the worker, if everything we compute will be used?

**Pull request merged**